hikari/impl/event_manager_base.py

```diff
@@ -50,46 +56,135 @@
 
     Specific event handlers should be in functions named `on_xxx` where `xxx`
     is the raw event name being dispatched in lower-case.
     """
 
-    __slots__: typing.Sequence[str] = ("_app", "_listeners", "_consumers", "_waiters")
+    __slots__: typing.Sequence[str] = ("_app", "_consumers", "_enabled_consumers_cache", "_listeners", "_waiters")
 
     def __init__(self, app: traits.BotAware) -> None:
         self._app = app
-        self._consumers: typing.Dict[str, ConsumerT] = {}
+        self._consumers: typing.Dict[str, _Consumer] = {}
+        self._enabled_consumers_cache: typing.Dict[_Consumer, bool] = {}
         self._listeners: ListenerMapT[base_events.Event] = {}
         self._waiters: WaiterMapT[base_events.Event] = {}
 
         for name, member in inspect.getmembers(self):
             if name.startswith("on_"):
-                self._consumers[name[3:]] = member
+                member = typing.cast(MethodT, member)
+                cache_resource = getattr(member, _CACHE_RESOURCE_ATTRIBUTE, undefined.UNDEFINED)
+                event_types = getattr(member, _EVENT_TYPES_ATTRIBUTE, undefined.UNDEFINED)
+                cache_resource: undefined.UndefinedOr[config.CacheComponents] = cache_resource
+                event_types: undefined.UndefinedOr[typing.Sequence[typing.Type[base_events.Event]]] = event_types
+                self._consumers[name[3:]] = _Consumer(member, cache_resource, event_types)
+
+    def _clear_enabled_cache(self) -> None:
+        self._enabled_consumers_cache = {}
+
+    def _enabled_for_event(self, event_type: typing.Type[base_events.Event], /) -> bool:
+        for cls in event_type.dispatches():
+            if cls in self._listeners or cls in self._waiters:
+                return True
+
+        return False
+
+    # This returns int rather than bool to avoid an unnecessary bool cast
+    def _enabled_for_consumer(self, consumer: _Consumer, /) -> int:
+        # If undefined then we can only assume that this may link to registered listeners.
+        if consumer.event_types is undefined.UNDEFINED:
+            return True
+
+        if (cached_value := self._enabled_consumers_cache.get(consumer)) is True:
+            return True
+
+        if cached_value is None:
+            for event_type in consumer.event_types:
+                if event_type in self._listeners or event_type in self._waiters:
+                    self._enabled_consumers_cache[consumer] = True
+                    return True
+
+            self._enabled_consumers_cache[consumer] = False
+
+        # If cache_components is UNDEFINED then we have to fall back to assuming that the consumer might set state.
+        # If cache_components is NONE then it doesn't make set state calls.
+        return (
+            consumer.cache_components is undefined.UNDEFINED
+            or consumer.cache_components != config.CacheComponents.NONE
+            and consumer.cache_components & self._app.cache.settings.components
```

Comment:
```suggestion
            and (consumer.cache_components & self._app.cache.settings.components) != 0
```

The fact that this returns `int` makes it really weird :confused: